helix-core/src/main/java/org/apache/helix/HelixManager.java

```diff
@@ -419,6 +420,17 @@ void addExternalViewChangeListener(org.apache.helix.ExternalViewChangeListener l
    */
   Long getSessionStartTime();
 
+  /**
+   * Checks whether the cluster manager is leader and returns the session ID associated to the
+   * connection of cluster data store, if and only if it is leader.
+   *
+   * @return {@code Optional<String>} session ID is present inside the {@code Optional} object
+   * if the cluster manager is leader. Otherwise, returns an empty {@code Optional} object.
+   */
+  default Optional<String> getSessionIdIfLead() {
```

Comment:
getSessionIdIfLeader?  Also, why do we need this if we already have manager.getSessionId?

Reply:
manager.getSessionIdIfLead checks whether the cluster manager is leader and returns the session ID associated to the connection of cluster data store, if and only if it is leader. It combines isLeader() and getSessionId() in one method to ensure the session id returning is what we check leadership in isLeader().

Without it, we may do this:

if (isLeader()) {
  sessionId = manager.getSessionId();
}
The problem is race condition. sessionId may change after isLeader().

So we need manager.getSessionIdIfLead to check leadership and get sessionId at the same time.

Reply:
Basically, the idea is that check leader and get leader sessionID at one operation, later if sessionID changes, we use sessionAwareWriter interface to fail it.

Reply:
@kaisun2000 You are right. This is the core idea.

Reply:
One more thing for `why do we need this if we already have manager.getSessionId?`:
`getSessionId` just directly returns the existing sessionId in manager. But `getSessionIdIfLeader()` does more thing by reading zk to check leadership. This is the main difference.